### DISCUSSION TITLE
Question: Is it possible to disable X propagation and treat Xs as zeroes?

### DISCUSSION THREAD

**Discussion by flaviens** [steveicarus/iverilog#1025] [Q&A] [answered]
Hi!

My apologies, the answer is probably somewhere but I could really not find it yet :crying_cat_face: .
**Is it possible to treat undefined values (X)  as zeroes, like Verilator would do?**

Thank you in advance!
Flavien

**Comment by martinwhitaker** [comment, marked as answer]
[moved to Discussions]

You would need to either cast them to a 2-state type or assign them to a 2-state variable.

**Reply by flaviens** [reply]
Thank you!

**Comment by caryr** [comment]
I agree with what Martin said, it sounds like you need to cast these to two state equivalents. We do not currently have an option to do this in the compiler, though you could easily write a script to alter the code. I doubt it would be that hard to add this to Icarus. Are you doing this for a functionality reason to match what verilator does or are you trying to get a faster simulation? In theory two state simulations should be faster, but that is in general not currently true in Icarus. Also note that even if you switch to two state variables there is at least one place where you will still get a four state result as mandated by the standard (division by zero). In most cases this would likely get mapped to 0/false eventually, but there are corner cases (e.g. `1/0 ? first : second` would not return the `second` variable as the result, but instead the mixing of the two).

I believe verilator is completely two state for both speed and simplicity and just blindly maps all operations to two-state so it may not handle this specific case correctly.

**Comment by flaviens** [comment]
Thank you for your answers! The goal is functionality indeed. I will keep an eye on what you said about divisions by zero. If you have it on top of your mind, do you have a link to a list of these corner cases? Thanks!

**Reply by caryr** [reply]
These corner cases are in the standard, but that can be a challenging place to find all the details. I believe things that look like division by zero are the only exception (i.e. division, modulus, negative power of a zero value).

**Reply by flaviens** [reply]
Ok thanks!
